Update Getting Started Docs: Updated environment variables and add dev env setup instructions (#867)
DataBiosphere/jade-data-repo: docs/jade-getting-started.md

# Getting Started

Ensure that you have access to the required team resources beforehand. If you
encounter a permission error, it is likely because you are missing appropriate
access. Join the `#github` Slack channel, click the lightning bolt in the
channel header, and select `Join DataBiosphere`. Ask for access to Google Groups
including `dsde-engineering`.

These instructions assume you use MacOS, and that you are on the internal Broad
network or the VPN. If the VPN is not installed, follow the instructions
[at this link](https://broad.io/vpn).

> During this process, you will need your GitHub and Docker Hub username,
password, and personal access token for multiple steps, so make sure to have
those handy. If you don't have those yet, see the section below, otherwise you
can skip to [Connect Accounts](#2-connect-accounts)

> 

Reviewer: What is the thought behind moving this section?

Author: No good reason - I've moved it back!

Reviewer: No strong concern either way, I just wanted to see if you had any thoughts on this. 👍

Create a [personal access token](https://help.github.com/en/github/authenticating-to-github/creating-a-personal-access-token-for-the-command-line)
so you can interact with GitHub on the command line.

During this process, you will need your GitHub and Docker Hub username,
password, and personal access token for multiple steps, so make sure to have
those handy.

> ## 2. Request Required Access

Reviewer: ```suggestion
## 2. Request Required Access

```

nit: add newline between markdown sections

Ensure that you have access to the required team resources. If you
encounter a permission error, it is likely because you are missing appropriate

> access. 

Reviewer: ```suggestion
access.

```

nit: add newline between markdown sections

- DataBiosphere: Join the `#github` Slack channel, click the lightning bolt in the
channel header, and select `Join DataBiosphere`.  Once you've been granted access
to DataBiosphere, ask a team member to add your github user to the
[DataBiosphere/jadeteam group](https://github.com/orgs/DataBiosphere/teams/jadeteam).
This will give you admin access to our repositories. 

> - Google Groups: Ask a team member for access to Google Groups including `dsde-engineering`.

Reviewer: ```suggestion
- Google Groups: Ask a team member for access to Google Groups including `jade-internal` and `dsde-engineering`.
```

then click in the search box and install **Cloud Code**, which integrates
Google Cloud features with IntelliJ IDEA.

## 6. Create GitHub token

The GitHub token verifies team permissions. This token is necessary for the next
step, [Login to Vault](#6-login-to-vault). To create a token:

1. Go to the [GitHub Personal Access Token](https://github.com/settings/tokens)
page and click **Generate new token**.
2. Give the token a descriptive name, **only** give it the following two scopes and then click **Generate token**.
  *  `read:org` scope under `admin:org`

>   * `workflow` (this will give you access to kick off github actions from the command line)

Author: @samanehsan  - I know you just ran into the error when you were trying to run ./test-dispatch.sh from the command line. Was this the solution for you?

Reviewer 2: My token has workflow and full repo access, which is probably too permissive

1. Follow [instructions in our terraform-jade repository](https://github.com/broadinstitute/terraform-jade/tree/master/old#new-team-member-process) to add your initials to the terraform templates and generate the static resources needed to deploy your personal development environment. Apply the changes and create a pull request to merge your additons to terraform-jade.
2. Create your datarepo helm definition:
  -  In `datarepo-helm-definitions/dev` directory, copy an existing developer
definition and change all initials to your own.
  -  Create a pull request with these changes in [datarepo-helm-definitions](https://github.com/broadinstitute/datarepo-helm-definitions)
3. Connect to your new dev postgres database instance (replace ZZ with your inintials):
Note that this is separate instance than the local one you will configure in step 9. The following command connected to the database via a proxy.
```
cd jade-data-repo/ops
DB=datarepo SUFFIX=ZZ ENVIRONMENT=dev ./db-connect.sh
```

> 4. Add some setting to your database, that I can't remembmer... 

Author: I just requested access to the recording on our session from Monday - I'll update this after watching that!

> 5. Ask colleague from DevOps to create a google project for your named broad-jade-ZZ (replacing 'ZZ' with your initials) in the broadinstitute.org organization

Author: @tcjiang9  - I added the step that you completed this afternoon!

git clone https://github.com/broadinstitute/datarepo-helm-definitions
```

> ## 8. [DRAFT] Set up your Development Environment

Reviewer: ```suggestion
## 8. Set up your Development Environment

```

Is this still a draft?

nit: add newline between markdown sections

The goal of this step is set up some of the basic components of your development
environment.  You'll actually spin up this instance on broad-jade-dev in next step.

> Note: These instructions have not been tested yet! This may be a good step to
> pair on with another jade team member. There is a video of us walking through
> these steps in our [Jade Googe Drive Folder](https://drive.google.com/drive/folders/1JM-_M0qsX6eXocyPc9TB7ivCKJTji3dX?usp=sharing). 

Reviewer: ```suggestion
> These instructions have not been tested yet! This may be a good step to
pair on with another Jade team member. There is a video of us walking through
these steps in our [Jade Google Drive Folder](https://drive.google.com/drive/folders/1JM-_M0qsX6eXocyPc9TB7ivCKJTji3dX?usp=sharing). 
```

> 1. Follow [instructions in our terraform-jade repository](https://github.com/broadinstitute/terraform-jade/tree/master/old#new-team-member-process)

Reviewer: ```suggestion
1. Follow the [instructions in our terraform-jade repository](https://github.com/broadinstitute/terraform-jade/tree/master/old#new-team-member-process)
```

to add your initials to the terraform templates and generate the static resources needed
to deploy your personal development environment. Apply the changes and create a pull request

> to merge your additons to terraform-jade.

Reviewer: ```suggestion
to merge your additions to `terraform-jade`.
```

2. Create your datarepo helm definition:
  -  In `datarepo-helm-definitions/dev` directory, copy an existing developer
definition and change all initials to your own.
  -  Create a pull request with these changes in [datarepo-helm-definitions](https://github.com/broadinstitute/datarepo-helm-definitions)

> 3. Connect to your new dev postgres database instance (replace ZZ with your inintials):

Reviewer: ```suggestion
3. Connect to your new dev postgres database instance (replace `zz` with your initials):
```

Note that this is separate instance than the local one you will configure in step 9.

> The following command connects to the database via a proxy.

Reviewer: ```suggestion
The following command connects to the database via a proxy.

```

nit: add newline between markdown sections

```
cd jade-data-repo/ops

> DB=datarepo SUFFIX=ZZ ENVIRONMENT=dev ./db-connect.sh

Reviewer: ```suggestion
DB=datarepo SUFFIX=zz ENVIRONMENT=dev ./db-connect.sh
```

Author: I switched everything to ZZ

```

> 4. Now that you're connected to your dev database, run the following command

Reviewer: ```suggestion

4. Now that you're connected to your dev database, run the following command
```

nit: add newline between markdown sections

> (Once DR-1156 is done, this will no longer be needed):

Reviewer: ```suggestion
(Once DR-1156 is done, this will no longer be needed):

```

nit: add newline between markdown sections

```
create extension pgcrypto;
```

> 5. Ask a colleague from DevOps to create a google project for you with the following details:

Reviewer: ```suggestion

5. Ask a colleague from DevOps to create a google project for you with the following details:
```

nit: add newline between markdown sections

>   * Google Project Name: broad-jade-ZZ (replacing 'ZZ' with your initials)

Reviewer: ```suggestion
  * Google Project Name: `broad-jade-zz` (replacing `zz` with your initials)
```

1. Follow the [instructions in our terraform-jade repository](https://github.com/broadinstitute/terraform-jade/tree/master/old#new-team-member-process)
to add your initials to the terraform templates and generate the static resources needed
to deploy your personal development environment. Apply the changes and create a pull request
to merge your additions to `terraform-jade`.
2. Create your datarepo helm definition:
  -  In `datarepo-helm-definitions/dev` directory, copy an existing developer
definition and change all initials to your own.
  -  Create a pull request with these changes in [datarepo-helm-definitions](https://github.com/broadinstitute/datarepo-helm-definitions)
3. Connect to your new dev postgres database instance (replace `ZZ` with your initials):
Note that this is separate instance than the local one you will configure in step 9.
The following command connects to the database via a proxy.

```
cd jade-data-repo/ops
DB=datarepo SUFFIX=ZZ ENVIRONMENT=dev ./db-connect.sh
```

4. Now that you're connected to your dev database, run the following command

> (Once DR-1156 is done, this will no longer be needed):

Reviewer: ```suggestion
(Once DR-1156 is done, this will no longer be needed):
```

Should we link out to this item?

* Certain environment variables need to be set beforehand. Instances of `zzz`
should be replaced by your initials or the environment (i.e. `dev`):

```
export NVM_DIR="$HOME/.nvm"

> export JAVA_HOME=$(/usr/libexec/java_home)

Reviewer: Now that we are front loading these variables, this list seems overwhelming imo. A number of these variables seem unnecessary. I think we should only include the absolute minimum required to run integration tests.

Author: Good point - I think I've now narrowed it down to just the important ones. I added a note about converting between application.properties and environment variables, which should cover any settings I left out.

export VAULT_ADDR=https://clotho.broadinstitute.org:8200
export ENVIRONMENT=dev

> export GOOGLE_CLOUD_PROJECT=broad-jade-zzz

Reviewer: ```suggestion
export GOOGLE_CLOUD_PROJECT=broad-jade-zzz
```

We should either switch everything to `zz` or keep it as `zzz`

Author: Good point! I've switched everything to `ZZ`.

export GOOGLE_CLOUD_DATA_PROJECT=broad-jade-zzz-data
export SUFFIX=zzz
export PROXY_URL=https://jade-zzz.datarepo-dev.broadinstitute.org
export CYPRESS_BASE_URL=http://local.broadinstitute.org:3000
export GOOGLE_APPLICATION_CREDENTIALS=/tmp/jade-dev-account.json
export GOOGLE_SA_CERT=/tmp/jade-dev-account.pem
export JADECLI_TEST_BILLING_ACCOUNT=00708C-45D19D-27AAFA
export DATAREPO_USEREMAIL={your dev gmail account}

> # Clears database on startup, test run, etc. (Further explaned in oncall playbook)

Reviewer: ```suggestion
# Clears database on startup, test run, etc. This is further explained in the oncall playbook.
```

export DB_MIGRATE_DROPALLONSTART=true

# Setting for testing environment (Further explaned in oncall playbook)
export GOOGLE_ALLOWREUSEEXISTINGBUCKETS=true
export GOOGLE_ALLOWREUSEEXISTINGPROJECTS=true

> # Integration test setting: change this to http://localhost:8080/ to run against  # local instance

Reviewer: ```suggestion
# Integration test setting: change this to http://localhost:8080/ to run against a local instance
```

* You will need to run `render-configs.sh` before running integration tests.

> * **Set Environment Variables**: While not exhaustive, here's a list that notes the important environment variables to set when running jade-data-repo locally. Instances of `ZZ` should be replaced by your initials or the environment (i.e. `dev`).  These variables override settings in

Reviewer: ```suggestion
* **Set Environment Variables**: While not exhaustive, here's a list that notes the important environment variables to set when running `jade-data-repo` locally. Instances of `ZZ` should be replaced by your initials or the environment (i.e. `dev`).  These variables override settings in
```

  1. Go to the [Google Cloud Console](http://console.cloud.google.com/).
  2. From the `DATA.TEST-TERRA.BIO` organization, select your newly created GCP
  project: `broad-jade-zzz-data`.

>   3. Go to Firestore and enable firestore in **Native** mode.

Author: Should no longer be required w/ Muscles' fix!

### 2. Build `jade-data-repo-ui`

Follow the [setup instructions](https://github.com/DataBiosphere/jade-data-repo-ui#jade-data-repository-ui)
to build the `jade-data-repo-ui` repository.

### 3. Initialize `terraform-jade`

Follow the [setup instructions](https://github.com/broadinstitute/terraform-jade#terraform-jade)
to initialize the `terraform-jade` repository.

> 

Author: This is now a part of step 8. Not totally sure the best place for it to live.
